client/src/actors/secure.rs

```diff
@@ -842,277 +827,260 @@ impl Handler<CallProcedure> for SecureClient {
+    }
+    let key = self.keystore.take_key(msg.vault_id).unwrap();
 
     let mut seed = vec![0u8; msg.size_bytes];
-    fill(&mut seed).map_err(|e| anyhow::anyhow!(e))?;
+    match fill(&mut seed) {
+        Ok(_) => {},
+        Err(e) => {
+            self.keystore.insert_key(msg.vault_id, key);
+            return Err(anyhow::anyhow!(e))
+        }
+    }
 
-    match self.db.write(&key, msg.vault_id, msg.record_id,&seed, msg.hint).map_err(|e| anyhow::anyhow!(e)) {
-        Ok(_) => {
-            Ok(crate::ProcResult::SLIP10Generate(StatusMessage::OK))
-        },
-        Err(e) => Err(anyhow::anyhow!(e))
+    let res = self.db.write(&key, msg.vault_id, msg.record_id, &seed, msg.hint);
+
+    self.keystore.insert_key(msg.vault_id, key);
+
+    match res {
+        Ok(_) => Ok(ProcResult::SLIP10Generate(StatusMessage::OK)),
+        Err(e) => Err(anyhow::anyhow!(e)),
     }
 });
 
 impl_handler!(procedures::SLIP10DeriveFromSeed, Result<crate::ProcResult, anyhow::Error>, (self, msg, _ctx), {
-    match self.keystore.get_key(msg.seed_vault_id) {
-        Some(seed_key) => {
-            self.keystore.insert_key(msg.seed_vault_id, seed_key.clone());
-            let dk_key = if !self.keystore.vault_exists(msg.key_vault_id) {
-                let key = self.keystore.create_key(msg.key_vault_id);
-                self.db.init_vault(&key, msg.key_vault_id).map_err(|e| anyhow::anyhow!(e))?;
-                key
-
-            } else {
-                self.keystore.get_key(msg.key_vault_id).ok_or_else(||
-                    Err::<crate::ProcResult, anyhow::Error>(anyhow::anyhow!(crate::Error::KeyStoreError("".into())))
-                ).unwrap()
-            };
-
-            self.keystore.insert_key(msg.key_vault_id, dk_key.clone());
-
-            // FIXME if you see this fix here, that a single-threaded mutable reference
-            // is being passed into the closure to obtain the result of the pro-
-            // cedure calculation, you should consider rethinking this approach.
-
-            let result = Rc::new(Cell::default());
-
-            match self.db.exec_proc(&seed_key, msg.seed_vault_id, msg.seed_record_id, &dk_key, msg.key_vault_id, msg.key_record_id, msg.hint, |gdata| {
-                let dk = Seed::from_bytes(&gdata.borrow())
-                    .derive(Curve::Ed25519, &msg.chain).map_err(|e| anyhow::anyhow!(e)).unwrap();
-                let data : Vec<u8> = dk.into();
-
-                // was formerly sent to the client
-                result.set(dk.chain_code());
-
-                Ok(data)
-            }) {
-                Ok(_) => {
-                    let result = result.get();
-                    Ok(ProcResult::SLIP10Derive(ResultMessage::Ok(result)))
-                }
-                Err(e) => {
-                    Err(anyhow::anyhow!(e))
-                }
+    let seed_key = self
+        .keystore
+        .take_key(msg.seed_vault_id)
+        .ok_or(anyhow::anyhow!(VaultError::NotExisting))?;
```

Comment:
Seems like a very round-about way to return an error.  Might be better to use `map_err` in the `take_key` function and then just return the error directly that way with 
```
let seed_key = self
        .keystore
        .take_key(msg.seed_vault_id)?
```

Would require changing the `take_key` return to an `Result` type but might be useful given how many times you do this conversion.   See how often its the `Option` type is being used in the context of the rest of the code to weigh whether or not this change is worth it overall.